### PR TITLE
PortBridge - Reliability updates

### PR DESCRIPTION
**Pull request by dlstucki**
## Description
- Fix a data corruption issue where InputQueue ended up enqueuing items out of order. Integrate the Inputqueue.cs changes from here: https://referencesource.microsoft.com/#System.ServiceModel.Internals/System/Runtime/InputQueue.cs
- Ensure to read the whole preamble to avoid reacting to a partially read "frame" preamble.
- Throw if any read returns 0 bytes when framing protocol indicates there should be more.
- Log Listener Online, Reconnecting, and Offline events.
- Remove two unused classes/files.
- Trace textual error description from Enum instead of decimal socket error code.
- If TcpClientConnectionForwarder is not open (has been closed) there's no need to attempt another BeginAcceptTcpClient call.
- Remove an uncalled .ctor from ThrottlingQueueBufferedStream which could easily lead to null-refs since it didn't assign any value to the semaphore field.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] If applicable, the code is properly documented.
- [x] The code builds without any errors.

### CODE REVIEW

File: samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs [hunk 1]
```diff
@@ -66,9 +67,14 @@ public void DoneReadingPreamble(IAsyncResult readOutputAsyncResult)
...
-                    if (bytesRead == 1)
+                    while (bytesRead < preambleBuffer.Length)
                     {
-                        bytesRead += bufferRead(preambleBuffer, 1, preambleBuffer.Length - 1);
+                        int newBytesRead = this.bufferRead(preambleBuffer, bytesRead, preambleBuffer.Length - bytesRead);
+                        bytesRead += newBytesRead;
+                        if (newBytesRead == 0)
+                        {
+                            throw new ProtocolViolationException("Unexpected EOF while reading preamble");
+                        }
```
Comment (bainian12345) [inline samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs]: this can go in a separate method?

**Comment by bainian12345** [comment]
>             preambleBuffer = new byte[sizeof (int) + sizeof (ushort)];

I would make these into a constant so it's easier to understand what they are for

---
Refers to: samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs:35 in 8a001f6. [](commit_id = 8a001f645e19c505f6529fa3fedb0adc98264f59, deletion_comment = False)

**Comment by dlstucki** [comment]
>             preambleBuffer = new byte[sizeof (int) + sizeof (ushort)];

I would argue that's outside of the scope of this fix.

---
In reply to: [574852205](https://github.com/Azure/azure-relay/pull/50#issuecomment-574852205) [](ancestors = 574852205)

---
Refers to: samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs:35 in 8a001f6. [](commit_id = 8a001f645e19c505f6529fa3fedb0adc98264f59, deletion_comment = False)

**Comment by bainian12345** [comment]
>         public void Open()

shouldn't `Open()` and `Close()` use the lock that `Dispose()` uses?

---
Refers to: samples/hybrid-connections/dotnet/portbridge/PortBridge/TcpClientConnectionForwarder.cs:54 in 8a001f6. [](commit_id = 8a001f645e19c505f6529fa3fedb0adc98264f59, deletion_comment = False)

**Comment by dlstucki** [comment]
>             preambleBuffer = new byte[sizeof (int) + sizeof (ushort)];

I added a comment. In my mind "sizeof(int)" is basically a constant and I wouldn't want to alias that intrinsic value.

---
In reply to: [574855064](https://github.com/Azure/azure-relay/pull/50#issuecomment-574855064) [](ancestors = 574855064,574852205)

---
Refers to: samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs:35 in 8a001f6. [](commit_id = 8a001f645e19c505f6529fa3fedb0adc98264f59, deletion_comment = False)

File: samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs (hunk 1 above)
Comment (dlstucki) [reply samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs]: Done. Let me know if you have a better method name. I'm not totally sold on mine.

---
In reply to: [367089552](https://github.com/Azure/azure-relay/pull/50#discussion_r367089552) [](ancestors = 367089552)

**Comment by dlstucki** [comment]
>         public void Open()

The lock is for the keeping the connections collection consistent. `open` is not directly related to that. I made it volatile and removed it from the lock statement the one place it was in there.

---
In reply to: [574855426](https://github.com/Azure/azure-relay/pull/50#issuecomment-574855426) [](ancestors = 574855426)

---
Refers to: samples/hybrid-connections/dotnet/portbridge/PortBridge/TcpClientConnectionForwarder.cs:54 in 8a001f6. [](commit_id = 8a001f645e19c505f6529fa3fedb0adc98264f59, deletion_comment = False)

File: samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs (hunk 1 above)
Comment (bainian12345) [reply samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs]: `assertNotEndOfStream` maybe? I was actually thinking the whole block could go in a method.

---
In reply to: [367151813](https://github.com/Azure/azure-relay/pull/50#discussion_r367151813) [](ancestors = 367151813,367089552)

Comment (dlstucki) [reply samples/hybrid-connections/dotnet/portbridge/PortBridge/MultiplexConnectionInputPump.cs]: Ah, ok. Done in iteration 3.

---
In reply to: [367155549](https://github.com/Azure/azure-relay/pull/50#discussion_r367155549) [](ancestors = 367155549,367151813,367089552)

**Pull request merged**